ddtrace/tracer/option.go

```diff
@@ -138,6 +138,26 @@ type config struct {
 
 	// enabled reports whether tracing is enabled.
 	enabled bool
+
+	// postProcessor holds the function used to process finished spans of a trace.
+	// It reports whether the trace should be dropped.
+	postProcessor func([]ddtrace.ReadWriteSpan) (dropTrace bool)
+}
+
+// WithPostProcessor sets post processor function f, which allows certain span fields to change
+// after a trace has finished, and can force a trace to be dropped despite any existing sampling
+// rules. Modifying the span can be done by using methods of the ReadWriteSpan interface, such
+// as SetTag. If f returns true, the entire trace will be dropped. If f returns false, then no
+// changes to the sampling method will be made based on the post processor function, though the
+// trace may be dropped in the agent due to other sampling rules.
+//
+// Please note that dropped traces and spans will not be accounted for in APM stats, and that
+// modifying any of the ReadWriteSpans outside the scope of the processor function can and will
+// likely cause a panic.
+func WithPostProcessor(f func([]ddtrace.ReadWriteSpan) (dropTrace bool)) StartOption {
```

Comment:
Had discussion with @ajgajg1134 and @knusbaum about this. Wanted to share the proposal for our design change:

The new API:

```
// WithOnFinish sets function f to run after the span has finished, which allows certain span fields to change before being sent to the Datadog Agent, or be abandoned entirely.
// Any spans that are not present in the returned `[]ddtrace.ReadWriteSpan` will not be accounted for in APM stats, or be sent to the Datadog Agent.
// Please note that modifying any of the ReadWriteSpans outside the scope of the processor function can and will likely cause a panic.
func WithOnFinish(f func([]ddtrace.ReadWriteSpan) ([]ddtrace.ReadWriteSpan)) StartOption
```

It's similar to your existing API, so the code changes should hopefully be minimal, but the benefits of this are that it makes the API a bit clearer, and it allows individual spans to be abandoned in the trace.

Apologies for not catching this during RFC review. This is something we are going to reflect on as a team to see how we can improve RFC reviews in the future in hopes of gathering these ideas before implementation has begun.

Reply:
Sounds good, this makes sense to me! In that case, the user can return nil or an empty slice to indicate that the full trace needs to be dropped.

Working on this 🙂

Reply:
Exactly. That would be good to call out specifically in the godoc as well 👍  Thanks!!